docs/DeveloperGuide.md

```diff
@@ -1,34 +1,216 @@
+## 1. Design
+### Architecture
+The Duke class is the main class of the product. It is responsible for:
+* At app launch: Initializes the components in the correct sequence, and connects them up with each other.
+* At shut down: Shuts down the components and invokes storage method where necessary.	
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+The product also contains the following components:
+* Cards: Holds the data in the type of Card and relative operations
+* Subjects: Holds the data in the type of Subject and relative operations
+* Score: Holds the data in the type of Score and relative operations
+* Commands: Deals with user input and communicate CLI to relative methods
+* Exceptions: Deals with illegal inputs
+* Parser: Convert CLI inputs into command keywords
 
+In these components, cards and subjects have similar structure. Both of them contains a Card/Subject class and CardList/SubjectList class.
+Duke, along with all command class and Parser form the logic box of the product.
+<br />![](images/logicuml.jpg)
 
-## Product Scope
-### Target user profile
+<br />Logic box interacts with Model box, i.e. Card, CardList, Subject, SubjectList, ScoreList.
+<br />![](images/modeluml.jpg)
 
-{Describe the target user profile}
+<br />Finally the Storage box, i.e. Storage class will handle reading and writing the content to files.
+The Storage component saves the SubjectList objects in Serializable format and loads it back.
+<br />![](images/storageuml.jpg)
 
-### Value proposition
+## 2. Implementation
+### 2.1. [Proposed] Subject Feature
+#### 2.1.1. Proposed Implementation
+The subject feature is an extension to the existing flashcard feature which allows users to categorise their
+flashcards. This helps users to search for their flashcards more efficiently and also users to quiz by subject. The list of user's subjects are stored inside the SubjectList. It implements the following operations:
 
-{Describe the value proposition: what problem does it solve?}
+- ``SubjectList#addSubject()`` - Adds a new subject to the subject list.
+- ``SubjectList#removeSubject()`` - Removes an existing subject from the subject list.
+- ``SubjectList#listSubjects()`` - List the subjects in the subject list.
 
-## User Stories
+Step 1. Before the user decides to add a flashcard, he/she can create a subject to store the flashcard using
+the command ``addsubject s/SUBJECTNAME``.
+The following diagram describes how the add subject operation works:
+
+![](images/addsubject_sequence_uml.jpg)
+
+Step 2. The user executes the command ``listsubjects`` to view the subjects currently stored in the application.
+
+Step 3. Once the user has chosen a subject, he/she can execute the command ``addcard s/SUBJECTINDEX q/QUESTION a/ANSWER``
+to add a flashcard into the subject. 
+The following diagram describes how the add card operation works:
+
+![](images/addcard_sequence_uml.jpg)
```

Comment:
Should there be an arrow here for the return of the cardList?
![image](https://user-images.githubusercontent.com/42372756/77991014-56915180-7355-11ea-9674-69a7a6a68123.png)

Reply:
Also, be careful of the arrow. It should hit the edge of the box.

![image](https://user-images.githubusercontent.com/59989652/77993297-724b2680-735a-11ea-8221-e8a580bb906f.png)

Reply:
Also,  The dotted line seems to be inconsistent. You can compare the difference.

![image](https://user-images.githubusercontent.com/59989652/77993559-f3a2b900-735a-11ea-989a-6a1baf116dd9.png)

![image](https://user-images.githubusercontent.com/59989652/77993625-13d27800-735b-11ea-86d3-1ef2d5e916fd.png)